src/Nethermind/Nethermind.Trie/NibbleExtensions.cs

```diff
@@ -18,26 +21,46 @@ public static Nibble[] FromBytes(params byte[] bytes)
         public static Nibble[] FromBytes(ReadOnlySpan<byte> bytes)
         {
             Nibble[] nibbles = new Nibble[2 * bytes.Length];
-            for (int i = 0; i < bytes.Length; i++)
-            {
-                nibbles[i * 2] = new Nibble((byte)((bytes[i] & 240) >> 4));
-                nibbles[i * 2 + 1] = new Nibble((byte)(bytes[i] & 15));
-            }
-
+            BytesToNibbleBytes(bytes, MemoryMarshal.AsBytes(nibbles.AsSpan()));
             return nibbles;
         }
 
-        public static void BytesToNibbleBytes(ReadOnlySpan<byte> bytes, Span<byte> nibbles)
+        public unsafe static void BytesToNibbleBytes(ReadOnlySpan<byte> bytes, Span<byte> nibbles)
         {
             if (nibbles.Length != 2 * bytes.Length)
             {
                 ThrowArgumentException();
             }
 
-            for (int i = 0; i < bytes.Length; i++)
+            var length = bytes.Length / sizeof(Vector128<byte>) * sizeof(Vector128<byte>);
+            if (Vector128.IsHardwareAccelerated && length > 0)
+            {
+                var input = MemoryMarshal.Cast<byte, Vector128<byte>>(bytes.Slice(0, length));
+                ref var output = ref Unsafe.As<byte, Vector128<ushort>>(ref MemoryMarshal.GetReference(nibbles));
+
+                for (int i = 0; i < input.Length; i++)
+                {
+                    Vector128<byte> value = input[i];
+                    (Vector128<ushort> lower0, Vector128<ushort> upper0) = Vector128.Widen(Vector128.BitwiseAnd(value, Vector128.Create((byte)0xf)));
+                    lower0 = Vector128.Shuffle(lower0.AsByte(), Vector128.Create((byte)1, 0, 1, 2, 1, 4, 1, 6, 1, 8, 1, 10, 1, 12, 1, 14)).AsUInt16();
+                    upper0 = Vector128.Shuffle(upper0.AsByte(), Vector128.Create((byte)1, 0, 1, 2, 1, 4, 1, 6, 1, 8, 1, 10, 1, 12, 1, 14)).AsUInt16();
+                    (Vector128<ushort> lower1, Vector128<ushort> upper1) = Vector128.Widen(Vector128.BitwiseAnd(value, Vector128.Create((byte)0xf0)));
+                    lower1 = Vector128.ShiftRightLogical(lower1.AsByte(), 4).AsUInt16();
+                    upper1 = Vector128.ShiftRightLogical(upper1.AsByte(), 4).AsUInt16();
+
+                    lower0 = Vector128.BitwiseOr(lower0, lower1);
+                    upper0 = Vector128.BitwiseOr(upper0, upper1);
+
+                    Unsafe.Add(ref output, i * 2) = lower0;
+                    Unsafe.Add(ref output, i * 2 + 1) = upper0;
+                }
+            }
+
+            for (int i = length; i < bytes.Length; i++)
             {
-                nibbles[i * 2] = (byte)((bytes[i] & 240) >> 4);
-                nibbles[i * 2 + 1] = (byte)(bytes[i] & 15);
+                int value = Unsafe.Add(ref MemoryMarshal.GetReference(bytes), i);
+                Unsafe.Add(ref MemoryMarshal.GetReference(nibbles), i * 2) = (byte)(value >> 4);
```

Comment:
is this actually faster than plain array access?

Reply:
The array access has all sorts of extra bounds checking stuff on it because its a weird access; and also doesn't start at 0, so the Jit can't eliminate them; so end up with 3 checks -> exceptions; this change has no checks; and logically we have verified all the numbers up top (that nibbles is double size and we only go to length)